### ISSUE TITLE
add a function to assert Uniqueness

### ISSUE THREAD

**Issue by fmms** [Bungalow64/DBConfirm#96]
Hi,

thanks for this very nice project.

I have just played a bit and was expecting a function to confirm uniqueness. i.e.

```csharp
  [TestMethod]
       public async Task unique()
        {
            QueryResult data1 = await TestRunner.ExecuteCommandAsync("SELECT 1 a, 2 b UNION SELECT 2, 2 ORDER BY a");
            data1
              .AssertColumnCount(2)
              .AssertColumnExists("a")
              .AssertColumnsNotExist("c", "d")
              .AssertValue(0, "a", 1) 
// missing?
              .AssertColumnUniqe("a")
              .AssertColumnsUniqe("a", "b")
            ;
        }
```

Did I miss something? Could those be added?

regards
Felix

**Comment by Bungalow64** [comment]
Hey Felix,

Thanks for raising this - I know exactly what you mean, let me check the API again to make sure we've not got something that can already cover this, otherwise I'll have a think about how best to add this.

**Comment by Bungalow64** [comment]
So you're right, the API doesn't really allow you to test against duplicate column names.

I've tried with a couple of tests to check with and without duplicates:

```csharp
[Test]
public async Task ReturnDuplicateColumnName_AssertNoDuplicate()
{
    QueryResult data = await TestRunner.ExecuteCommandAsync("SELECT Orders.OrderID, Orders.OrderDate FROM Orders");

    data
        .AssertColumnCount(2)
        .AssertColumnExists("OrderID")
        .AssertColumnExists("OrderDate")
        .AssertColumnsExist("OrderID", "OrderDate") // Passes, as expected
        .AssertColumnsExist("OrderID", "OrderID", "OrderDate"); // Currently passes, maybe should fail?
}

[Test]
public async Task ReturnDuplicateColumnName_AssertDuplicate()
{
    QueryResult data = await TestRunner.ExecuteCommandAsync("SELECT Orders.OrderID, Orders.OrderID, Orders.OrderDate FROM Orders");

    data
        .AssertColumnCount(3)
        .AssertColumnExists("OrderID")
        .AssertColumnExists("OrderDate")
        .AssertColumnsExist("OrderID", "OrderDate") // Passes, as expected
        .AssertColumnsExist("OrderID", "OrderID", "OrderDate"); // Also passes, as expected
}
```

To verify that there are no duplicates, we can get around this by checking the column count (AssertColumnCount) and checking that each expected column exists (using AssertColumnsExist).  If there's a duplicate, then either the count won't match, or if the count does match then there must be a different column missing.  But this isn't ideal, because it won't indicate which column has been duplicated.

And to verify that there are duplicates, again we'd have to rely on AssertColumnCount and AssertColumnsExist, but there's no way to check which column specifically has been duplicated.

I'll do a bit more testing and see what we'd need to add to the API to cover each of these.

**Comment by Bungalow64** [comment]
Ah, ok, I see.  So you want to verify that all the values in a particular column are unique?  So if a duplicate value is found, you want the test to fail.

Say you get these results back:

| a  | b |
| ------------- | ------------- |
| 1  | 2  |
| 2  | 2  |

If you called `.AssertColumnValuesUnique("a")` then that would pass because all the values in column "a" are unique.
But if you called `.AssertColumnValuesUnique("b")` then that would fail because there's a duplicate value in "b".
And if we had something like `.AssertColumnValuesUnique("a", "b")` then that would also fail because of the duplicate in "b".

Is that what you're looking for?

**Comment by fmms** [comment]
@Bungalow64 i would assume you a,b example to pass, because then it is a composite key. With your example the combination of both columns is unique for every row.

**Comment by Bungalow64** [comment]
Yep, that makes perfect sense, thanks.  I'll try and get a PR up for this in the next couple of weeks, unless you want to have a go at it, in which case feel free to assign this issue to yourself.

**Comment by jamie-burns** [comment]
This has been merged into the develop branch, for v1.1.0.  This issue will be updated once the changes have been deployed to NuGet.
